nx/lib/io/nx_io.ml

```diff
@@ -100,6 +100,162 @@ module Safe = struct
+          let rec next_nonempty () =
+            try
+              let line = input_line ic in
+              if String.trim line = "" then next_nonempty () else Some line
+            with End_of_file -> None
+          in
+          let split_line line =
+            let line = String.trim line in
+            if sep = "" then [ line ]
+            else if String.length sep = 1 then
+              List.filter (fun s -> s <> "")
+                (String.split_on_char sep.[0] line)
+            else (
+              (* split by substring 'sep' and drop empty parts *)
+              let rec aux acc start =
+                if start >= String.length line then List.rev acc
+                else
+                  match String.index_from_opt line start sep.[0] with
+                  | None ->
+                      let part = String.sub line start (String.length line - start) in
+                      let acc = if part = "" then acc else part :: acc in
+                      List.rev acc
+                  | Some idx ->
+                      if idx + String.length sep <= String.length line
+                         && String.sub line idx (String.length sep) = sep
+                      then
+                        let part = String.sub line start (idx - start) in
+                        let acc = if part = "" then acc else part :: acc in
+                        aux acc (idx + String.length sep)
+                      else aux acc (idx + 1)
+              in
+              aux [] 0)
+          in
+          match next_nonempty () with
+          | None -> Error (Format_error "Empty file")
+          | Some first_line ->
+              let first_vals = split_line first_line in
+              let cols = List.length first_vals in
+              if cols = 0 then Error (Format_error "No columns detected")
+              else (
+                let rows_rev = ref [ first_vals ] in
+                let rec loop () =
+                  match next_nonempty () with
+                  | None -> ()
+                  | Some line ->
+                      let vals = split_line line in
+                      if List.length vals <> cols then
+                        raise (Failure "Inconsistent number of columns")
+                      else (
+                        rows_rev := vals :: !rows_rev;
+                        loop ())
+                in
+                loop ();
+                let rows = List.rev !rows_rev in
+                let rows_count = List.length rows in
+                (* Read as float64 then cast to requested dtype *)
+                let ba =
+                  Bigarray_ext.Genarray.create Bigarray_ext.Float64 Bigarray.c_layout
+                    [| rows_count; cols |]
+                in
+                List.iteri
+                  (fun i vals ->
+                    List.iteri
+                      (fun j v ->
+                        Bigarray_ext.Genarray.set ba [| i; j |]
+                          (float_of_string v))
```

Comment:
We parse every token via float_of_string before casting. Large int64s (>|2^53|) will get mangled at that point, so integer round-trips fail. We need to parse integers directly (Int64.of_string, etc.) based on the target dtype.